rtsp.go

```diff
@@ -658,11 +672,39 @@ func newRTSPCamera(ctx context.Context, _ resource.Dependencies, conf resource.C
 	}
 	cancelCtx, cancel := context.WithCancel(context.Background())
 	reader := gostream.VideoReaderFunc(func(_ context.Context) (image.Image, func(), error) {
-		latest := rc.latestFrame.Load()
-		if latest == nil {
+		if videoCodec(rc.currentCodec.Load()) == MJPEG {
+			img := rc.latestMJPEGImage.Load()
+			if img == nil {
+				return nil, func() {}, errors.New("no mjpeg frame yet")
+			}
+			return *img, func() {}, nil
+		}
+
+		rc.frameSwapMu.Lock()
+		defer rc.frameSwapMu.Unlock()
+
+		if rc.latestFrame == nil {
 			return nil, func() {}, errors.New("no frame yet")
 		}
-		return *latest, func() {}, nil
+		// Makes sure we use the same frame pointer in release as we do in this VideoReaderFunc.
+		frame := rc.latestFrame
+
+		frame.mu.Lock()
+		defer frame.mu.Unlock()
+
+		release := func() {
+			// When the caller is done with the image, we return the AVFrame to the pool such
+			// that we can re-use its allocated byte buffer.
+			frame.mu.Lock()
+			defer frame.mu.Unlock()
+
+			if refCount := frame.decrementRefs(); refCount == 0 {
+				rc.avFramePool.put(frame)
+			}
+		}
+
+		frame.incrementRefs()
```

Comment:
Because I left a comment directly on the comment, GH is forbidding me from submitting a review. So I'll say, let's move the `incrementRefs` up closer to the `frame := rc.latestFrame` assignment. That's when the `VideoReaderFunc` takes a stake in the lifetime of the frame. And it makes things more clear that the `frameSwapMu` is intended to be locked around that.

Reply:
Got it. Done